lib/remote_node.go

```diff
@@ -0,0 +1,468 @@
+		connectionType:         RemoteNodeConnectionType_Unknown,
+		connectionStatus:       RemoteNodeConnectionStatus_NotConnected,
+		handshakeMetadata:      nil,
+		srv:                    srv,
+		bc:                     bc,
+		cmgr:                   cmgr,
+		minTxFeeRateNanosPerKB: minTxFeeRateNanosPerKB,
+		hyperSync:              hyperSync,
+		posValidator:           posValidator,
+	}
+}
+
+// SetConnectedPeer sets the peer for the remote node, and updates the remote node's connection type,
+// connection status, and ID.
+func (rn *RemoteNode) SetConnectedPeer(peer *Peer) {
+	rn.peer = peer
+	if rn.peer == nil {
+		return
+	}
+
+	// Set connectionType
+	if rn.peer.IsOutbound() {
+		rn.connectionType = RemoteNodeConnectionType_Outbound
+		rn.setId(NewRemoteNodeOutboundId(peer.GetId(), peer.GetAttemptId()))
+	} else {
+		rn.connectionType = RemoteNodeConnectionType_Inboound
+		rn.setId(NewRemoteNodeInboundId(peer.GetId()))
+	}
+
+	// Set connectionStatus
+	rn.connectionStatus = RemoteNodeConnectionStatus_Connected
+}
+
+func (rn *RemoteNode) GetPeer() *Peer {
+	return rn.peer
+}
+
+func (rn *RemoteNode) IsInbound() bool {
+	return rn.peer != nil && !rn.peer.IsOutbound()
+}
+
+func (rn *RemoteNode) IsOutbound() bool {
+	return rn.peer != nil && rn.peer.IsOutbound()
+}
+
+func (rn *RemoteNode) setId(id RemoteNodeId) {
+	rn.id = id
+}
+
+func (rn *RemoteNode) GetId() RemoteNodeId {
+	return rn.id
+}
+
+func (rn *RemoteNode) GetPeerId() uint64 {
+	peerId, _ := rn.id.GetIds()
+	return peerId
+}
+
+func (rn *RemoteNode) CreateOutboundConnection(netAddr *wire.NetAddress) {
+	if rn.connectionStatus != RemoteNodeConnectionStatus_NotConnected {
+		return
+	}
+
+	attemptId := rn.cmgr.DialOutboundConnection(netAddr)
+	id := NewRemoteNodeAttemptedId(attemptId)
+	rn.setId(id)
+	rn.connectionStatus = RemoteNodeConnectionStatus_Attempted
+}
+
+func (rn *RemoteNode) CreatePersistentOutboundConnection(netAddr *wire.NetAddress) {
+	if rn.connectionStatus != RemoteNodeConnectionStatus_NotConnected {
+		return
+	}
+
+	attemptId := rn.cmgr.DialPersistentOutboundConnection(netAddr)
+	id := NewRemoteNodeAttemptedId(attemptId)
+	rn.setId(id)
+	rn.connectionStatus = RemoteNodeConnectionStatus_Attempted
+}
+
+// ConnectInboundPeer connects a peer once a successful inbound connection has been established.
+func (rn *RemoteNode) ConnectInboundPeer(conn net.Conn, na *wire.NetAddress) error {
+	peer := rn.cmgr.ConnectPeer(conn, na, RemoteNodeIdNoAttempt, false, false)
+	if peer == nil {
+		return errors.Errorf("ConnectInboundPeer: Problem connecting peer (%s)", conn.RemoteAddr().String())
+	}
+	rn.SetConnectedPeer(peer)
+	return nil
+}
+
+// ConnectOutboundPeer connects a peer once a successful outbound connection has been established.
+func (rn *RemoteNode) ConnectOutboundPeer(conn net.Conn, na *wire.NetAddress, attemptId uint64, isOutbound bool, isPersistent bool) error {
+	peer := rn.cmgr.ConnectPeer(conn, na, attemptId, isOutbound, isPersistent)
```

Comment:
should we update `ConnectPeer` to actually return us an error? Can do this in a separate workstream, but seems like we're just guessing why a connection failed and may lead to headaches later on.

Reply:
Actually, it looks like `cmgr.ConnectPeer` should never fail so I just turned `ConnectOutboundPeer` and `ConnectInboundPeer` into voids.

Reply:
Hmm, well maybe it's still worthwhile for these functions to return an error if RemoteNode `connectionStatus` is incorrect.